testing/btest/Files/ws/wstest.py

```diff
@@ -0,0 +1,225 @@
+import os
+import time
+from typing import Any, Callable, Optional, Union
+
+import websockets.sync.client
+from websockets.sync.client import ClientConnection
+
+WS_PORT = (
+    int(os.environ["WEBSOCKET_PORT"].split("/")[0])
+    if "WEBSOCKET_PORT" in os.environ
+    else 0
+)
+
+# IPv4 non-secure WebSocker URL for version 1
+WS4_URL_V1 = f"ws://127.0.0.1:{WS_PORT}/v1/messages/json"
+
+DEFAULT_RECV_TIMEOUT = 0.1
+OWN_TOPIC_PREFIX = "/zeek/wstest"
+
+MAIN_TRIES = 100
+MAIN_SLEEP = 0.1
+
+
+class TestClient:
+    """
+    Helper class wrapping a websockets ClientConnection
+    with a bit of convenience.
+    """
+
+    def __init__(self, name: str, cc: ClientConnection):
+        self.__name = name
+        self.__cc = cc
+        self.__own_topic = f"{OWN_TOPIC_PREFIX}/{self.name}/"
+
+    def __enter__(self) -> "TestClient":
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback) -> None:
+        self.__cc.__exit__(exc_type, exc_value, traceback)
+
+    def hello_v1(self, topics: list[str]):
+        self.send_json([self.__own_topic] + topics[:])
+        ack = self.recv_json()
+        assert "type" in ack, repr(ack)
+        assert ack["type"] == "ack"
+        assert "endpoint" in ack, repr(ack)
+        assert "version" in ack
+        return ack
+
+    def recv_json(self, timeout=DEFAULT_RECV_TIMEOUT) -> dict:
+        raw = self.__cc.recv(timeout=timeout)
+        return json.loads(raw)
+
+    def send_json(self, data) -> None:
+        return self.__cc.send(json.dumps(data))
+
+    @property
+    def name(self) -> str:
+        return self.__name
+
+
+def connect(name: str, url: Optional[str] = None) -> TestClient:
+    """
+    Connect to a WebSocket server and return a TestClient instance.
+    """
+    if url is None:
+        url = WS4_URL_V1
+
+    cc = websockets.sync.client.connect(url)
+    return TestClient(name, cc)
+
+
+def recv_until_timeout(
+    clients: list[TestClient], *, timeout: float = DEFAULT_RECV_TIMEOUT, desc=None
+):
+    """
+    Read events from all entries in clients, stopping when all timeout,
+    otherwise print status messages.
+    """
+    if desc:
+        print(f"recv_until_timeout: {desc}")
+
+    all_timeout = False
+    while not all_timeout:
+        msgs = []
+        all_timeout = True
+        for tc in clients:
+            data = None
+            try:
+                data = tc.recv_json(timeout=timeout)
+                ev = data["data"][2]["data"]
+                info = {
+                    "client": tc.name,
+                    "topic": data["topic"],
+                    "event_name": ev[0]["data"],
+                    "event_args": ev[1]["data"],
+                }
+                msgs += [info]
+                all_timeout = False
+            except TimeoutError:
+                msgs += [{"client": tc.name, "timeout": True}]
+
+        if not all_timeout:
+            for msg in msgs:
+                print(json.dumps(msg))
+
+
+class TaggedArg:
```

Comment:
```suggestion
class TypedArg:
```
(totally up to you, but avoids wondering what the tagging is about)